### PR TITLE
[WiP] Support Linux platforms in Minikube bootstrapping script

### PR DESCRIPTION
**Pull request by alnikyur**
Steps to prepare installation minikube bootstrapping on Linux OS
==========================
- Check if minikube, kubectl and helm is installing
- Check versions of those binary ( diff hash of current binary with latest hash of binary on site )
- Run minikube locally on Linux with vm-driver=none

### CODE REVIEW

**Comment by tyranron** [comment]
@alnikyur please, explain in the description the purpose of this PR. And clarify then why you've assigned it to me.

**Comment by alnikyur** [comment]
@tyranron test.sh includes next functions:
1. checkos - define distribution of linux os (rhel or debian like). I made this function to work with different package managers if we needed this feature in future.
2. checkInstalledSoft - define minikube, helm and kubectl on machine, if not installed - install (using binary package instead package manager).
3. updatePackages - update minikube, helm and kubectl if their versions are lower then:
kubectlVer="1.9.1"
minikubeVer="0.24.1"
helmVer="2.8.0"
Need to discuss what version is normally to use.
4. checkVerionTest - analog of updatePackages, but not alreade done. Need to discuss what function is better.
I made functions in separate file for better understanding what i was done.
I assigned this PR to you accidentally.

**Comment by tyranron** [comment]
@alnikyur why are you implementing that in separate script? Just extend original `bootstrap.sh` script. And move info into PR's description. It must state clearly what it does and why.

File: minikube/bootstrap.sh [hunk 1]
```diff
@@ -116,6 +116,128 @@ waitDashboardIsDeployed() {
   set -e
 }
 
+path="/usr/local/bin/"
```
Comment (tyranron) [inline minikube/bootstrap.sh]: @alnikyur this should not be defined statically "somewhere between functions", but rather be configurable env var with default `/usr/local/bin/` value.

Define it appropriate section `# Execution` where other env vars are defined.

File: minikube/bootstrap.sh [hunk 2]
```diff
@@ -116,6 +116,128 @@ waitDashboardIsDeployed() {
...
+
+# initVersion init current and latest stable version
+initVersion(){
+
+minikubeStableVersion=$( curl -s https://github.com/kubernetes/minikube/releases/ \
```
Comment (tyranron) [inline minikube/bootstrap.sh]: @alnikyur why there is no indentation in your bash functions? It's too hard to read!

File: minikube/bootstrap.sh [hunk 3]
```diff
@@ -140,6 +262,10 @@ case "$OS" in
     # TODO: Hyperkit driver is still not stable enough. Use with later releases.
     #installHyperkitDriver
     ;;
+  linux)
+    checkIfInstall
+    upgradeLinuxSoft
```
Comment (tyranron) [inline minikube/bootstrap.sh]: @alnikyur it's better not to split this actions, but rather do install-or-upgrade action for each software at once.

File: minikube/bootstrap.sh [hunk 4]
```diff
@@ -116,6 +116,128 @@ waitDashboardIsDeployed() {
...
+}
+
+# upgradeLinuxSoft Check if the latest version of minikube, kubectl and helm is installed
+# if not - upgrade to latest stable version
+upgradeLinuxSoft(){
+
+initVersion
+
+if [[ "$minikubeCurrentVersion" == "$minikubeStableVersion" ]]; then
+  echo "Minikube latest version"
```
Comment (tyranron) [inline minikube/bootstrap.sh]: @alnikyur no sense in such `echo`s. Better use `runCmd` (it has `set -x` inside).

**Review by tyranron — CHANGES_REQUESTED** [review]
The initial idea is not "to write you own script inside existing one", but rather "to extend existing script to support Linux platform".

That's why you implementation is too rude:
- The code style of existing script is not followed well.
- You should use `set -x` capabilities already present in script (to follow the implemented idea), not playing with custom `echo`s.

Additionally:
- Think about better way to download latest binaries. Your current link parsing is too complicated and fragile. It easily can stop working on some HTML changes. That is not the way.

**Comment by alnikyur** [comment]
Rebuild bootstrap.sh
================
Add next functions:
- helmLinuxUpgrade
- kubectlLinuxUpgrade
- minikubeLinuxUpgrade
- helmLinuxInstall
- kubectlLinuxInstall
- minikubeLinuxInstall
- upgradeLinuxPackages

Annotations of functions bellow in comments.
Add two variants of define and compare versions (please see  Add linux support (#1) and  Add linux support (#2) commits)

File: minikube/bootstrap.sh [hunk 5]
```diff
@@ -116,147 +116,106 @@ waitDashboardIsDeployed() {
...
-    /tmp/minikube.sha && cat /tmp/minikube.sha`
-
-CURRENT_VER=`sha256sum /usr/local/bin/minikube | awk '{print $1}'`
-
-if ! [[ "$LATEST_VER_SHA" == "CURRENT_VER" ]]; then
-echo "curl -s -LO https://github.com/kubernetes/minikube/releases/download/v$VER/minikube-linux-$ARCH"
-fi
+#helmLinuxUpgrade check new version and upgrade helm using version number
+helmLinuxUpgrade() {
+  unset LATEST_VER CURRENT_VER SHA CHECK_VER
```
Comment (tyranron) [inline minikube/bootstrap.sh]: Use `local varName="value"` to not hacking with `unset`.

File: minikube/bootstrap.sh [hunk 6]
```diff
@@ -116,147 +116,106 @@ waitDashboardIsDeployed() {
...
+  LATEST_VER=`curl -s https://github.com/kubernetes/helm/releases/latest | sed -e 's/.*v\(.*\)".*/\1/'`
```
Comment (tyranron) [inline minikube/bootstrap.sh]: Use `https://api.github.com/repos/kubernetes/helm/releases/latest` to deal with JSON structured data, not HTML.

See [example](https://github.com/instrumentisto/gitlab-builder-docker-image/blob/980045487eea0aadb930ecdfea0666b8fe6b9aa7/Dockerfile#L46-L50).

File: minikube/bootstrap.sh [hunk 7]
```diff
@@ -116,147 +116,106 @@ waitDashboardIsDeployed() {
...
+  CURRENT_VER=`helm version 2>/dev/null | sed -n '/.*v/s///p' | grep -o '^[^"]*' | head -n 1`
+  if ! [[ "$CURRENT_VER" == "$LATEST_VER" ]]; then
```
Comment (tyranron) [inline minikube/bootstrap.sh]: Nope, it's better to compare file hashes. Because files may change even for the same version.

Comment (tyranron) [reply minikube/bootstrap.sh]: Oh, I see it's [here](https://github.com/instrumentisto/toolchain/pull/2/files/2031a80ece1bc1fa4fd28214b3619a1976659d65..3b7e9c5a5bef4e275ec299e9906a365e35950f0d#diff-904c401c51ca9723b75a72547bdb4f63R133).

Why bother to implement 2 funcs? Leave only variant with hashes.

**Review by tyranron — CHANGES_REQUESTED** [review]
Please, do described changes.

File: minikube/bootstrap.sh (hunk 7 above)
Comment (alnikyur) [reply minikube/bootstrap.sh]: @tyranron I wrote this variant as alternative and want to ask you what is better. I`ll use variant with hashes.

**Comment by tyranron** [comment]
@alnikyur why did you close this?

**Comment by alnikyur** [comment]
@tyranron This PR was deleted automatically after renamed from `minikube-bootstrap-fro-linux` to `minikube-bootstrap-for-linux` - didn't know, thati if branch was renamed - PR automatically removed.

**Comment by tyranron** [comment]
@alnikyur ah, OK.

**Pull request closed without merging**